File: app/src/main/java/com/raival/fileexplorer/tabs/file/tasks/CompressTask.java
package com.raival.fileexplorer.tabs.file.tasks;

import android.os.Handler;
import android.os.Looper;

import com.raival.fileexplorer.tabs.file.model.Task;
import com.raival.fileexplorer.utils.ZipUtils;

import java.io.File;
import java.util.ArrayList;

public class CompressTask extends Task {
    private final ArrayList<File> filesToCompress;
    private File zipFile;

    public CompressTask(ArrayList<File> filesToCompress) {
        this.filesToCompress = filesToCompress;
    }

    public ArrayList<File> getFilesToCompress() {
        return filesToCompress;
    }

    @Override
    public String getName() {
        return "Compress";
    }

    @Override
    public String getDetails() {
        StringBuilder sb = new StringBuilder();
        boolean first = true;

        for (File file : filesToCompress) {
            if (!first) {
                sb.append(", ");
            }
            sb.append(file.getName());
            first = false;
        }
        return sb.toString();
    }

    @Override
    public boolean isValid() {
        for (File file : filesToCompress) {
            if (!file.exists())
                return false;
        }
        return true;
    }

    @Override
    public void setActiveDirectory(File directory) {
        zipFile = directory;
    }

    @Override
    public void start(OnUpdateListener onUpdateListener, OnFinishListener onFinishListener) {
        new Thread(() -> {
            new Handler(Looper.getMainLooper()).post(() -> onUpdateListener.onUpdate("Compressing...."));
            try {
                ZipUtils.archive(filesToCompress, zipFile);
                new Handler(Looper.getMainLooper()).post(() -> onFinishListener.onFinish("Files have been compressed successfully"));
            } catch (Exception e) {
                e.printStackTrace();
                new Handler(Looper.getMainLooper()).post(() -> onFinishListener.onFinish(e.toString()));
            }
        }).start();
    }
}

File: app/src/main/java/com/raival/fileexplorer/tab/file/task/CompressTask.kt
package com.raival.fileexplorer.tab.file.task

import com.raival.fileexplorer.tab.file.misc.archive
import com.raival.fileexplorer.tab.file.model.Task
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File

class CompressTask(private val filesToCompress: ArrayList<File>) : Task() {
    private var zipFile: File? = null
    override val name: String
        get() = "Compress"
    override val details: String
        get() {
            val sb = StringBuilder()
            var first = true
            for (file in filesToCompress) {
                if (!first) {
                    sb.append(", ")
                }
                sb.append(file.name)
                first = false
            }
            return sb.toString()
        }
    override val isValid: Boolean
        get() {
            for (file in filesToCompress) {
                if (!file.exists()) return false
            }
            return true
        }

    override fun setActiveDirectory(directory: File) {
        zipFile = directory
    }

    override fun start(onUpdate: OnUpdateListener, onFinish: OnFinishListener) {
        CoroutineScope(Dispatchers.IO).launch {
            withContext(Dispatchers.Main) {
                onUpdate.onUpdate("Compressing....")
            }
            try {
                archive(filesToCompress, zipFile)
                withContext(Dispatchers.Main) {
                    onFinish.onFinish("Files have been compressed successfully")
                }
            } catch (e: Exception) {
                e.printStackTrace()
                withContext(Dispatchers.Main) {
                    onFinish.onFinish(e.toString())
                }
            }
        }
    }
}